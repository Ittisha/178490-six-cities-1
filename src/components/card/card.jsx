import * as React from 'react';
import PropTypes from 'prop-types';
import {Link} from 'react-router-dom';

import {getRatingPercent} from '../../utils/get-rating-percent';
import {
  ApartmentTypes,
  SmallCardImage,
  DefaultCardImage,
  SmallBookmarkButton,
} from '../../consts';
import BookmarkButton from '../bookmark-button/bookmark-button';
import apartmentPropsShape from '../../props/apartment';

class Card extends React.PureComponent {
  constructor(props) {
    super(props);

    this._handleImgClick = this._handleImgClick.bind(this);
  }

  render() {
    const {apartment, cardClass, imageWrapperClass, infoClass, isSmall} = this.props;
    const {
      isPremium,
      isInBookmarks,
      price,
      rating,
      photoUrl,
      title,
      type,
      id
    } = apartment;

    const imgWidth = isSmall ? SmallCardImage.WIDTH : DefaultCardImage.WIDTH;
    const imgHeight = isSmall ? SmallCardImage.HEIGHT : DefaultCardImage.HEIGHT;

    return (
      <article className={`place-card ${cardClass}`}>
        {isPremium && (
          <div className="place-card__mark">
            <span>Premium</span>
          </div>
        )}
        <div className={`place-card__image-wrapper ${imageWrapperClass}`}>
          <a href="#" onClick={this._handleImgClick}>
            <img className="place-card__image" src={photoUrl} width={imgWidth} height={imgHeight} alt="Place image"/>
          </a>
        </div>
        <div className={`place-card__info ${infoClass}`}>
          <div className="place-card__price-wrapper">
            <div className="place-card__price">
              <b className="place-card__price-value">&euro;{price}</b>
              <span className="place-card__price-text">&#47;&nbsp;night</span>
            </div>
            <BookmarkButton
              offerId={id}
              isInBookmarks={isInBookmarks}
              className="place-card"
              width={SmallBookmarkButton.WIDTH}
              height={SmallBookmarkButton.HEIGHT}
            />
          </div>
          <div className="place-card__rating rating">
            <div className="place-card__stars rating__stars">
              <span style={{width: `${getRatingPercent(rating)}`}}></span>
              <span className="visually-hidden">Rating</span>
            </div>
          </div>
          <h2 className="place-card__name">
            <Link to={`/offer/${id}`}>{title}</Link>
          </h2>
          <p className="place-card__type">{ApartmentTypes[type.toUpperCase()]}</p>
        </div>
      </article>);
  }

  _handleImgClick() {
    const {apartment, onImgClick} = this.props;
    if (typeof onImgClick === `function`) {
      onImgClick(apartment);
    }
  }
}

Card.propTypes = {
  apartment: apartmentPropsShape.isRequired,
  onImgClick: PropTypes.func,
  cardClass: PropTypes.string.isRequired,
  infoClass: PropTypes.string,
  imageWrapperClass: PropTypes.string.isRequired,
  isSmall: PropTypes.bool,
};

Card.defaultProps = {
  infoClass: ``,
  isSmall: false,
};

export {Card};
