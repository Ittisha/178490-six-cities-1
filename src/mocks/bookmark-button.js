export const MOCK_PROPS = {
  offerId: 1,
  isInBookmarks: false,
  className: ``,
  width: 15,
  height: 12,
};
