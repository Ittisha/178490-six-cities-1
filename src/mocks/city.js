export const INIT_CITY = {
  name: `Amsterdam`,
  coords: [52.38333, 4.9],
  zoom: 12,
};

export const MOCK_CITIES = [`Amsterdam`, `Paris`];

export const MOCK_CITIES_COORDS = {
  Amsterdam: [52.38333, 4.9],
  Paris: [48.864716, 2.349014],
};

export const MOCK_CITIES_ZOOM = {
  Amsterdam: 17,
  Paris: 15,
};
