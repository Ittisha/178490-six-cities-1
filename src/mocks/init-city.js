export const INIT_CITY = {
  name: `Amsterdam`,
  coords: [52.38333, 4.9],
};
