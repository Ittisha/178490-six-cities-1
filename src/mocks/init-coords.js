export const INIT_COORDS = [52.38333, 4.9];
