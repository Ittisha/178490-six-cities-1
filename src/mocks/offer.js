export const OFFER = {
  "city": {
    "name": `Cologne`,
    "location": {
      "latitude": 50.938361,
      "longitude": 6.959974,
      "zoom": 13}},
  "preview_image": `https://es31-server.appspot.com/six-cities/static/hotel/17.jpg`,
  "images": [
    `https://es31-server.appspot.com/six-cities/static/hotel/20.jpg`,
    `https://es31-server.appspot.com/six-cities/static/hotel/2.jpg`,
  ],
  "title": `The house among olive `,
  "is_favorite": false,
  "is_premium": false,
  "rating": 2.5,
  "type": `house`,
  "bedrooms": 2,
  "max_adults": 6,
  "price": 757,
  "goods": [
    `Coffee machine`,
    `Baby seat`,
    `Dishwasher`,
    `Fridge`,
    `Washing machine`,
    `Breakfast`,
  ],
  "host": {
    "id": 25,
    "name": `Angelina`,
    "is_pro": true,
    "avatar_url": `img/avatar-angelina.jpg`
  },
  "description": `This is a place for dreamers to reset, reflect, and create. Designed with a 'slow' pace in mind, our hope is that you enjoy every part of your stay; from making local coffee by drip in the morning, choosing the perfect record to put on as the sun sets.`,
  "location": {
    "latitude": 50.945361,
    "longitude": 6.935974,
    "zoom": 16
  },
  "id": 1
};
