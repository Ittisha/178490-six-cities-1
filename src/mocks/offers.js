export const OFFERS = [
  {
    isPremium: true,
    isInBookmarks: false,
    title: `Beautiful & luxurious apartment at great location`,
    photoUrl: `img/apartment-01.jpg`,
    price: 120,
    type: `Apartment`,
    rating: `93%`,
    id: 1,
    coordinates: [52.3909553943508, 4.85309666406198],
  },
  {
    isPremium: false,
    isInBookmarks: true,
    title: `Wood and stone place`,
    photoUrl: `img/room.jpg`,
    price: 80,
    type: `Private room`,
    rating: `80%`,
    id: 2,
    coordinates: [52.369553943508, 4.85309666406198],
  },
  {
    isPremium: false,
    isInBookmarks: false,
    title: `Canal View Prinsengracht`,
    photoUrl: `img/apartment-02.jpg`,
    price: 132,
    type: `Apartment`,
    rating: `80%`,
    id: 3,
    coordinates: [52.3909553943508, 4.929309666406198],
  },
  {
    isPremium: false,
    isInBookmarks: false,
    title: `Nice, cozy, warm big bed apartment`,
    photoUrl: `img/apartment-03.jpg`,
    price: 180,
    type: `Apartment`,
    rating: `80%`,
    id: 4,
    coordinates: [52.3809553943508, 4.939309666406198],
  }
];
