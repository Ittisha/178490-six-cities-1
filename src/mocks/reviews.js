export const MOCK_REVIEWS = [
  {
    id: 1,
    user: {
      avatarUrl: ``,
      id: 1,
      isPro: false,
      name: ``,
    },
    rating: 4,
    comment: ``,
    date: ``,
  },
  {
    id: 2,
    user: {
      avatarUrl: ``,
      id: 5,
      isPro: false,
      name: ``,
    },
    rating: 3,
    comment: ``,
    date: ``,
  }
];

export const MOCK_REVIEW = {
  comment: ``,
  id: 1,
  rating: `3`,
};
