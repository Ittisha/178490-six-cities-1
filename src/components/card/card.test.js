import * as React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import {StaticRouter} from "react-router-dom";
import {Card} from './card';
import {OFFERS} from '../../mocks/offers';

const mockApartment = OFFERS[0];

it(`Card component is rendered properly`, () => {
  const renderer = new ShallowRenderer();
  const tree = renderer
    .render(<StaticRouter>
      <Card
        apartment={mockApartment}
        onImgClick={() => {}}
        cardClass={`near-places__card`}
      />
    </StaticRouter>);

  expect(tree).toMatchSnapshot();
});
