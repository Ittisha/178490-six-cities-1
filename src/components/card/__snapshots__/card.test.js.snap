// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component is rendered properly 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  staticContext={Object {}}
>
  <Card
    apartment={
      Object {
        "bedrooms": 2,
        "cityCoords": Array [
          52.38333,
          4.9,
        ],
        "cityName": "Amsterdam",
        "cityZoom": 15,
        "coordinates": Array [
          52.3909553943508,
          4.85309666406198,
        ],
        "description": "description",
        "goods": Array [
          "Something",
        ],
        "host": Object {
          "avatar": "apartment-01.jpg",
          "id": 2,
          "isPro": true,
          "name": "Name",
        },
        "id": 1,
        "images": Array [
          "img/apartment-01.jpg",
        ],
        "isInBookmarks": false,
        "isPremium": true,
        "maxAdults": 4,
        "photoUrl": "img/apartment-01.jpg",
        "price": 120,
        "rating": 4.65,
        "title": "Beautiful & luxurious apartment at great location",
        "type": "Apartment",
        "zoom": 16,
      }
    }
    cardClass="near-places__card"
    imageWrapperClass="near-places__image-wrapper"
    infoClass=""
    isSmall={false}
    onImgClick={[Function]}
  />
</Router>
`;
